lib/msal-core/test/ScopeSet.spec.ts

```diff
@@ -1,16 +1,205 @@
+            expect(clientConfigError.message).to.contain(ClientConfigurationErrorMessage.nonArrayScopes.desc);
+        });
+
+        it("should throw createEmptyScopesArrayError if scopes is an empty array", function() {
+            const scopes = [];
+            let clientConfigError;
+
+            try {
+                ScopeSet.validateInputScope(scopes);
+            } catch (e) {
+                clientConfigError = e;
+            }
+
+            expect(clientConfigError instanceof ClientConfigurationError).to.eq(true);
+            expect(clientConfigError.errorCode).to.equal(ClientConfigurationErrorMessage.emptyScopes.code);
+            expect(clientConfigError.message).to.contain(ClientConfigurationErrorMessage.emptyScopes.desc);
+        });
     });
 
-    it("tests trimAndConvertToLowerCase", () => {
-        const scope = " S1 ";
-        expect(ScopeSet.trimAndConvertToLowerCase(scope)).to.be.eq("s1");
-    })
-});
+    describe("getScopeFromState", function() {
+        it("should return an empty string if null is passed in as state", function(){
+            expect(ScopeSet.getScopeFromState(null)).to.eql("");
+        });
+    });
+
+    describe("appendScopes", function() {
+        it("should return null if no scopes are passed into the first argument", function() {
+            expect(ScopeSet.appendScopes(null, ["s1"])).to.eql(null);
+        });
+
+        it("should return extended scopes if scopes and scopesToAppend are passed in", function() {
+            expect(ScopeSet.appendScopes(["s1"], ["s2"])).to.eql(["s1", "s2"]);
+        });
+                
+        it("should return extended scopes that include all the scopes in original scopes array and scopesToAppend", function() {
+            expect(ScopeSet.appendScopes(["s1", "s2"], ["s3", "s4"])).to.eql(["s1", "s2", "s3", "s4"]);
+        });
+
+        it("should return original scopes if scopesToAppend is null", function() {
+            expect(ScopeSet.appendScopes(["s1"], null)).to.eql(["s1"]);
+        });
+        
+        it("should return original scopes if scopesToAppend is an empty array", function() {
+            expect(ScopeSet.appendScopes(["s1"], [])).to.eql(["s1"]);
+        });
+    });
+
+    describe("generateLoginScopes", function() {
+        it("should append openid and profile to scopes, remove clientId from scopes if original scopes include clientId", function() {
+            expect(ScopeSet.generateLoginScopes([clientId], clientId)).to.include(openid);
+            expect(ScopeSet.generateLoginScopes([clientId], clientId)).to.include(profile);
+            expect(ScopeSet.generateLoginScopes([clientId], clientId)).to.not.include(clientId);
+        });
+
+        it("should append openid to scopes if original scopes are login scopes and does not include openid", function() {
+            expect(ScopeSet.generateLoginScopes([clientId, profile], clientId)).to.include(openid);
```

Comment:
Can we also verify profile is still there but clientId is not?

Reply:
Done.